pennylane/ops/qubit/hamiltonian.py

```diff
@@ -616,13 +616,30 @@ def compare(self, other):
         if isinstance(other, Hamiltonian):
             self.simplify()
             other.simplify()
-            return self._obs_data() == other._obs_data()  # pylint: disable=protected-access
+            if len(self._obs_data()) != len(other._obs_data()):
+                return False
+            print(self._obs_data(), " and ", other._obs_data())
+            for self_data in self._obs_data():  # pylint: disable=protected-access
+                is_equal = False
+                for other_data in other._obs_data():  # pylint: disable=protected-access
+                    if (
+                        qml.math.isclose(self_data[0], other_data[0])
+                        and self_data[1] == other_data[1]
+                    ):
+                        is_equal = True
+                        break
+                if not is_equal:
+                    return False
+            return True
 
         if isinstance(other, (Tensor, Observable)):
             self.simplify()
-            return self._obs_data() == {
-                (1, frozenset(other._obs_data()))  # pylint: disable=protected-access
-            }
+            if qml.math.isclose(list(self._obs_data())[0][0], 1) and list(self._obs_data())[0][
+                1
+            ] == frozenset(other._obs_data()):
+                return True
+            else:
+                return False
```

Comment:
```suggestion
            return qml.math.isclose(list(self._obs_data())[0][0], 1) and list(self._obs_data())[0][
                1
            ] == frozenset(other._obs_data())
```

Reply:
Actually directly returning the condition was causing some assertion errors in my device so I just rewrote it